web/src/engine/osk/src/input/gestures/specsForKeyboard.ts

```diff
@@ -0,0 +1,349 @@
+          //
+          // So, if the first tap of a key swaps layers, the second tap will be on the wrong layer and
+          // thus have a different state token.  This is the perfect place to detect and correct that.
+          if(comparisonSample.stateToken != incomingSample.stateToken) {
+            incomingSample.stateToken = comparisonSample.stateToken;
+
+            // TODO: specialized item lookup required here for proper 'correction', corresponding to
+            // the owning VisualKeyboard.  That rigging doesn't exist quite yet, at the time of writing this.
+            incomingSample.item = undefined;
+          }
+          return incomingSample.item == baseItem;
+        },
+      },
+      endOnResolve: true
+    }, {
+      model: InstantContactResolutionModel
+    }
+  ],
+  sustainTimer: {
+    duration: 500,
+    expectedResult: false,
+    baseItem: 'base'
+  },
+  resolutionAction: {
+    type: 'chain',
+    next: 'multitap',
+    item: 'current'
+  }
+}
+
+export const SimpleTapModel: GestureModel = {
+  id: 'simple-tap',
+  resolutionPriority: 1,
+  contacts: [
+    {
+      model: {
+        ...SimpleTapContactModel,
+        pathInheritance: 'chop',
+        itemPriority: 1
+      },
+      endOnResolve: true
+    }, {
+      model: InstantContactResolutionModel,
+      resetOnResolve: true
+    }
+  ],
+  resolutionAction: {
+    type: 'chain',
+    next: 'multitap',
+    item: 'current'
+  },
+  rejectionActions: {
+    item: {
+      type: 'replace',
+      replace: 'simple-tap'
+    }
+  }
+}
+
+export const SubkeySelectModel: GestureModel = {
+  id: 'subkey-select',
+  resolutionPriority: 0,
+  contacts: [
+    {
+      model: {
+        ...SubkeySelectContactModel,
+        pathInheritance: 'full',
+        itemPriority: 1
+      },
+      endOnResolve: true,
+      endOnReject: true
+    }, {
+      // A second touch while selecting a subkey will trigger instant cancellation
+      // of subkey mode.  (With this setting in place, anyway.)
+      //
+      // Might not be ideal for actual production... but it does have benefits for
+      // unit testing the gesture-matching engine.
+      model: InstantContactRejectionModel
+    }
+  ],
+  resolutionAction: {
+    type: 'complete',
+    item: 'current'
+  },
+  sustainWhenNested: true
+}
+
+export const ModipressStartModel: GestureModel = {
+  id: 'modipress-start',
+  resolutionPriority: 5,
+  contacts: [
+    {
+      model: {
+        ...ModipressContactStartModel,
+        allowsInitialState(incomingSample, comparisonSample, baseItem) {
+          // TODO:  needs better abstraction, probably.
+
+          // But, to get started... we can just use a simple hardcoded approach.
+          const modifierKeyIds = ['K_SHIFT', 'K_ALT', 'K_CTRL'];
+          for(const modKeyId of modifierKeyIds) {
+            if(baseItem.key.spec.id == modKeyId) {
+              return true;
+            }
+          }
```

Comment:
So modipress is only working for desktop modifier keys? That's going to need significant work still I think.

Reply:
Considering I've done little to actually integrate modipress thus far... yes, it will need significant work.